### PR TITLE
Outbound Foreign Keys

### PR DESCRIPTION
**Pull request by RFSH**
This PR adds the support for outbound foreign keys in reference.

Main changes are described in #203 (first comment). These are the lines that you can find them:
 - part 1 (#171): [changing the read call uri](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/reference.js#L563,L600), [attaching FKR values to page](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/reference.js#L1312,L1323)
 - part 2 ([Chaise Issue #654](https://github.com/informatics-isi-edu/chaise/issues/654)): [new contextualize](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/core.js#L1540,L1640), [PseudoColumn](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/core.js#L1894,L2015), [Reference.columns](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/reference.js#L263,L268)
 - part 3 (#159): [Tuple.values](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/reference.js#L1708,L1716), [PseudoColumn formatPresentation](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/core.js#L1994,L2010)

These are the other changes that were needed to be made in order to complete the task:
- Change the Reference constructor and `Reference.resolve` to be able to create Reference synchronously (#229).
- Add `InboundForeignKeys` class to represent inbound foreign keys and relocate the `_contextualize` to this class.
- Create `_generateRowName` function in utilities. Since it's being used in multiple places.

### CODE REVIEW

File: js/core.js [hunk 1]
```diff
@@ -1572,34 +1536,105 @@ var ERMrest = (function (module) {
...
+                            if (!colFound) continue;                            
+                            // avoid duplicate
+                            if (visiblePseudoColumns.indexOf(col) !== -1) continue;    
+                        }
 
-                    if (!columns.has(c.name)) { // not already in the columns.
-                        columns._push(c);
+                        visiblePseudoColumns.push(col);
+                        
+                    } catch (exception) {}
```
Comment (shinyichen) [inline js/core.js]: Empty exception handler?

File: js/core.js [hunk 2]
```diff
@@ -1853,6 +1890,129 @@ var ERMrest = (function (module) {
...
+        
+        this._foreignKey = foreignKey;
+        this._constraintName = foreignKey.constraint_names[0].join(":");
+        this._column = column;
+
+        /**
+         * @type {boolean}
+         * @desc indicator that this is a PseudoColumn rather than a Column.
+         */
+        this.isPseudo = true;
```
Comment (shinyichen) [inline js/core.js]: Add the flag to Column as well.

File: js/reference.js [hunk 3]
```diff
@@ -870,6 +947,7 @@ var ERMrest = (function(module) {
                     newRef = _referenceCopy(this);
                     delete newRef._context; // NOTE: related reference is not contextualized
                     delete newRef._related;
+                    delete newRef._pseudoColumns;
```
Comment (shinyichen) [inline js/reference.js]: Should this line be moved to `.setNewTable()` method?

Comment (shinyichen) [inline js/reference.js]: Also add this line to `.setNewTable()` method?

**Comment by RFSH** [comment]
I addressed the comments that @shinyichen and @jrchudy made. It's ready to merge.

Merging this will cause recordset to throw an error when user sorts using a `PseudoColumn`.  Therefore we should first merge [this PR in chaise](https://github.com/informatics-isi-edu/chaise/pull/686). 

However we should address it later and think about a better solution rather than just disabling (#246).

File: js/reference.js [hunk 4]
```diff
@@ -544,6 +564,62 @@ var ERMrest = (function(module) {
...
+                        tableIndex = 0,
+                        fkList = "",
+                        linking,
+                        sortColumn,
+                        keys,
+                        k;
+
+                    // add M alias to current table
+                    linking = parts[parts.length-1].match(/(\(.*\)=\(.*:.*:.*\))/);
+                    if (linking && linking[1]) { // the same logic as parser for finding the link syntax
```
Comment (shinyichen) [inline js/reference.js]: Just note to myself and @RFSH: Update this line to take search filter into account. See latest parser code.

**Pull request merged**